src/Sobol.jl

```diff
@@ -1,67 +1,101 @@
-        s = new(p)
-        finalizer(s, destroy)
-        s
-    end
+    m::Array{UInt32,2} #array of size (sdim, 32)
+    x::Array{UInt32,1} #previous x = x_n, array of length sdim
+    b::Array{UInt32,1} #position of fixed point in x[i] is after bit b[i]
+    n::UInt32 #number of x's generated so far
 end
-SobolSeq(N::Integer) = SobolSeq{N}()
-
-destroy(s::SobolSeq) = 
-    ccall((:nlopt_sobol_destroy,libnlopt), Void, (Ptr{Void},), s.p)
 
 ndims{N}(s::SobolSeq{N}) = N
 
-function show{N}(io::IO, s::SobolSeq{N})
-    print(io, "$N-dimensional Sobol sequence on [0,1]^$N")
-end
-function writemime{N}(io::IO, ::MIME"text/html", s::SobolSeq{N})
-    print(io, "$N-dimensional Sobol sequence on [0,1]<sup>$N</sup>")
-end
+function SobolSeq(N::Int)
+    (N < 0 || N > 1111) && error("invalid Sobol dimension")
+
+    m = ones(UInt32, (N, 32))
+
+    #special cases
+    N == 0 && return(SobolSeq{0})(m,UInt32[],UInt32[],zero(UInt32))
+    #special cases 1
+    N == 1 && return(SobolSeq{N}(m,UInt32[0],UInt32[0],zero(UInt32)))
 
-function skip!{N}(s::SobolSeq{N}, n::Integer, x::Vector{Float64})
-    (n < 0 || length(x) < N) && throw(BoundsError())
-    ccall((:nlopt_sobol_skip,libnlopt), Void, (Ptr{Void}, Cuint, Ptr{Float64}),
-          s.p, n, x)
-    x
+    for i = 2:N
+        a = sobol_a[i-1]
+        d = floor(Int, log2(a)) #degree of poly
+
+        #set initial values of m from table
+        # TODO transpose sobol_minit for faster column access?
+        m[i, 1:d] = sobol_minit[i-1, 1:d]
+        #fill in remaining values using recurrence
+        for j = (d+1):32
+            ac = a
+            m[i,j] = m[i,j-d]
+            for k = 0:d-1
+                m[i,j] $= ((ac & one(UInt32)) * m[i, j-d+k]) << (d-k)
+                ac >>= 1
+            end
+        end
+    end
+    SobolSeq{N}(m,zeros(UInt32,N),zeros(UInt32,N),zero(UInt32))
 end
 
-function skip{N}(s::SobolSeq{N}, n::Integer) 
-    skip!(s, n, Array(Float64, N))
-    nothing
+function next!{T<:FloatingPoint}(s::SobolSeq, x::AbstractVector{T})
+    length(x) != ndims(s) && throw(BoundsError())
+
+    if s.n == typemax(s.n)
+        rand!(x)
+        return nothing
+    end
+
+    s.n += 1
+    c = @compat UInt32(trailing_zeros(s.n))
+    for i=1:ndims(s)
+        b = s.b[i]
+        if b >= c
+            s.x[i] $= s.m[i,c+1] << (b-c)
+            x[i] = s.x[i] / (one(UInt32) << (b + one(UInt32)))
+        else
+            s.x[i] = (s.x[i] << (c-b)) $ s.m[i,c+1]
+            s.b[i] = c
+            x[i] = s.x[i] / (one(UInt32) << (c + one(UInt32)))
+        end
+    end
+    return x
 end
+next(s::SobolSeq) = next!(s, Array(Float64,ndims(s)))
 
-function next!{N}(s::SobolSeq{N}, x::Vector{Float64})
-    length(x) < N && throw(BoundsError())
-    ccall((:nlopt_sobol_next01,libnlopt), Void,
-          (Ptr{Void}, Ptr{Float64}), s.p, x)
-    x
+# Special case N == 0
+next(s::SobolSeq{0}) = Float64[]
+function next!{T<:FloatingPoint}(s::SobolSeq{0}, x::AbstractVector{T})
```

Comment:
Actually, now that I look at it, it seems like this may not be needed.  `next!{T<:FloatingPoint}(s::SobolSeq, x::AbstractVector{T})` should work fine for N = 0.